src/odm-collection.hpp

```diff
@@ -0,0 +1,656 @@
+    /// @throws exception::query if the operation fails.
+    ///
+    /// @see http://docs.mongodb.org/manual/core/read-operations-introduction/
+    ///
+    mongocxx::stdx::optional<T> find_one(
+        bsoncxx::document::view_or_value filter,
+        const mongocxx::options::find& options = mongocxx::options::find());
+
+    ///
+    /// Finds a single document matching the filter, deletes it, and returns the original as a
+    /// deserialized object.
+    ///
+    /// @param filter
+    ///   Document view representing a document that should be deleted.
+    /// @param options
+    ///   Optional arguments, see mongocxx::options::find_one_and_delete
+    ///
+    /// @return The deserialized object that was deleted from the database.
+    /// @throws exception::write if the operation fails.
+    ///
+    mongocxx::stdx::optional<T> find_one_and_delete(
+        bsoncxx::document::view_or_value filter,
+        const mongocxx::options::find_one_and_delete& options =
+            mongocxx::options::find_one_and_delete());
+
+    ///
+    /// Finds a single document matching the filter, replaces it, and returns either the original
+    /// or the replacement document as a deserialized object.
+    ///
+    /// @param filter
+    ///   Document view representing a document that should be replaced.
+    /// @param replacement
+    ///   Serializable object representing the replacement for a matching document.
+    /// @param options
+    ///   Optional arguments, see mongocxx::options::find_one_and_replace.
+    ///
+    /// @return The original or replaced object.
+    /// @throws exception::write if the operation fails.
+    ///
+    /// @note
+    ///   In order to pass a write concern to this, you must use the collection
+    ///   level set write concern - collection::write_concern(wc).
+    ///
+    mongocxx::stdx::optional<T> find_one_and_replace(
+        bsoncxx::document::view_or_value filter, const T& replacement,
+        const mongocxx::options::find_one_and_replace& options =
+            mongocxx::options::find_one_and_replace());
+
+    ///
+    /// Finds a single document matching the given serializable object, replaces it, and returns
+    /// either the original or the replacement document as a deserialized object.
+    ///
+    /// @param filter
+    ///   Serializable object representing a document that should be replaced.
+    /// @param replacement
+    ///   Serializable object representing the replacement for a matching document.
+    /// @param options
+    ///   Optional arguments, see mongocxx::options::find_one_and_replace.
+    ///
+    /// @return The original or replaced object.
+    /// @throws exception::write if the operation fails.
+    ///
+    /// @note
+    ///   In order to pass a write concern to this, you must use the collection
+    ///   level set write concern - collection::write_concern(wc).
+    ///
+    mongocxx::stdx::optional<T> find_one_and_replace(
```

Comment:
I made two versions, one that accepts a BSON document as a filter, and one that accepts a serializable object. I could see both use cases being useful, either filtering by some query, such as `{age: {$lt: 1}}`, or simply finding documents that exactly match a whole, serializable object.